apps/crl.c

```diff
@@ -231,7 +236,150 @@ int crl_main(int argc, char **argv)
+                nid = counts[i].nid;
+                cnt = counts[i].count;
+                sn = OBJ_nid2sn(nid);
+                if (sn == NULL)
+                    sn = OBJ_nid2ln(nid);
+                switch (nid) {
+                case NID_authority_key_identifier:
+                case NID_crl_number:
+                    if (cnt != 1) {
+                        BIO_printf(bio_err,
+                                   "CRL extension %s must appear "
+                                   "exactly once (found %d)\n",
+                                   sn, cnt);
+                        ret = 1;
+                        OPENSSL_free(counts);
+                        goto end;
+                    }
+                    break;
+                case NID_issuer_alt_name:
+                case NID_delta_crl:
+                case NID_issuing_distribution_point:
+                case NID_freshest_crl:
+                case NID_info_access:
+                    if (cnt > 1) {
+                        BIO_printf(bio_err,
+                                   "CRL extension %s must appear "
+                                   "at most once (found %d)\n",
+                                   sn, cnt);
+                        ret = 1;
+                        OPENSSL_free(counts);
+                        goto end;
+                    }
+                    break;
+                default:
+                    /* Any unknown critical extension is fatal */
+                    ext2 = NULL;
+                    for (k = 0; k < ext_total; k++) {
+                        tmp = sk_X509_EXTENSION_value(exts, k);
+                        if (OBJ_obj2nid(
+                                X509_EXTENSION_get_object(
+                                    tmp)) == nid) {
+                            ext2 = tmp;
+                            break;
+                        }
+                    }
+                    if (ext2 != NULL
+                        && X509_EXTENSION_get_critical(ext2)) {
+                        BIO_printf(bio_err,
+                                   "Unknown critical CRL "
+                                   "extension %s\n", sn);
+                        ret = 1;
+                        OPENSSL_free(counts);
+                        goto end;
+                    }
+                    break;
+                }
+            }
+            OPENSSL_free(counts);
+        }
+        /* Additional Authority Key Identifier semantic checks */
+        crit = 0;
+        idx = -1;
+        akid = X509_CRL_get_ext_d2i(x, NID_authority_key_identifier,
+                                    &crit, &idx);
+        if (akid) {
+            if (crit) {
+                BIO_printf(bio_err,
+                           "CRL extension Authority Key Identifier "
+                           "must not be marked critical\n");
+                AUTHORITY_KEYID_free(akid);
+                ret = 1;
+                goto end;
+            }
```

Comment:
I am opposed to this change. It serves no security purpose.  This is mere pedantic enforcement of interoperability requirements on issuing CAs that should not be enforced by CRL consumers.